sql-plugin/src/main/scala/com/nvidia/spark/rapids/ScalableTaskCompletion.scala

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util
+
+import org.apache.spark.TaskContext
+
+/**
+ * A handle that can be used to remove a callback if needed
+ */
+trait TaskCompletionCallbackHandle {
+  /**
+   * Removes the callback but does not call it if it has not already been called.
+   */
+  def removeCallback(): Unit
+
+  /**
+   * Removes the callback and calls it if it has not been called already.
+   */
+  def removeAndCall(): Unit
+}
+
+/**
+ * Provides a task completion listeners in Spark that you can remove if needed to help with scaling.
+ */
+object ScalableTaskCompletion {
+  // Note: There are three levels of synchronization that can happen here. The order that they are
+  // grabbed should always be
+  //
+  // ScalableTaskCompletion -> TopLevelTaskCompletion -> UserTaskCompletion
+  //
+  // to avoid any potential deadlocks
+
+  /**
+   * Wrapper around a user callback function.
+   */
+  private class UserTaskCompletion(val tc: TaskContext,
+      f: Either[() => Unit, TaskContext => Unit])
+      extends TaskCompletionCallbackHandle
+          with Function[TaskContext, Unit] {
+    private var wasCalled = false
+
+    override def removeCallback(): Unit = {
+      val topLevel = ScalableTaskCompletion.getTopLevel(tc.taskAttemptId())
+      if (topLevel != null) {
+        topLevel.remove(this)
+      }
+    }
+
+    override def apply(tc: TaskContext): Unit = synchronized {
+      if (!wasCalled) {
+        wasCalled = true
+        f match {
+          case Left(l) => l()
+          case Right(r) => r(tc)
+        }
+      }
+    }
+
+    override def removeAndCall(): Unit = {
+      removeCallback()
+      apply(tc)
+    }
+  }
+
+  /**
+   * Testing wrapper around the user callback function. It is not common and is here to allow
+   * us to still have things behave correctly in some unit tests.
+   */
+  private class TestTaskCompletion(f: () => Unit)
+      extends TaskCompletionCallbackHandle {
+    override def removeCallback(): Unit = {
+      // NOOP
+    }
+
+    override def removeAndCall(): Unit = f()
+  }
+
+  /**
+   * Keeps track of user callbacks for a given task. It is also the callback from
+   * Spark itself.
+   */
+  private class TopLevelTaskCompletion extends Function[TaskContext, Unit] {
+    private val pending = new util.Stack[UserTaskCompletion]()
```

Comment:
When I was running tests I had some debug logs and I saw at most about 7 or 8 entries during our unit tests.  That said in the real world I am not 100% sure how big it would be.